### PR TITLE
Add Unary mapping for bosonic operators

### PR DESCRIPTION
**Pull request by ddhawan11**
**Context:**
Added Unary mapping

**Description of the Change:**
Added functions to map bosonic operators to qubit form using unary mapping.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-72640]

### CODE REVIEW

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+from typing import Union
+from collections import defaultdict
+import numpy as np
+import pennylane as qml
+from pennylane.pauli import PauliSentence, PauliWord
+
+from .bosonic import BoseSentence, BoseWord
+
+
+def _get_pauli_op(i, j, qub_id):
```
Comment (soranjh) [inline pennylane/qchem/vibrational/bosonic_mapping.py:26]: Is this function needed for both binary and unary mapping or it is a duplicate from other PRs?

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+    c1, c2 = 0.5, -0.5 if i == 1 else 0.5
+
+    if i != j:
+        return PauliSentence({PauliWord({qub_id: "X"}): c1, PauliWord({qub_id: "Y"}): c2 * 1j})
+
+    return PauliSentence({PauliWord({}): c1, PauliWord({qub_id: "Z"}): c2})
+
+
+def binary_mapping(
```
Comment (soranjh) [inline pennylane/qchem/vibrational/bosonic_mapping.py:38]: Seems like a duplication that should not be here.

File: pennylane/qchem/vibrational/bosonic_mapping.py (hunk 1 above)
Comment (ddhawan11) [reply pennylane/qchem/vibrational/bosonic_mapping.py:26]: This function is needed for both mappings, so will take care of it when one of the PRs is merged.

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        nstates(int): maximum number of states a boson can occupy.
+        ps (bool): whether to return the result as a PauliSentence instead of an
+            operator. Defaults to False.
+        wire_map (dict): a dictionary defining how to map the states of
+            the Bose operator to qubit wires. If None, integers used to
+            label the bosonic states will be used as wire labels. Defaults to None.
+        tol (float): tolerance for discarding the imaginary part of the coefficients
+
+    Returns:
+        a linear combination of qubit operators
```
Comment (austingmhuang) [inline pennylane/qchem/vibrational/bosonic_mapping.py:62]: Same doc comments as binary mapping ones

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    **Example**
+
+    >>> w = qml.qchem.BoseWord()
+    >>> qml.qchem.unary_mapping(w, nstates=4)
```
Comment (austingmhuang) [inline pennylane/qchem/vibrational/bosonic_mapping.py:67]: Make sure to update these docs, forgot to mention in the other PRs.

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+def _(bose_operator: BoseWord, nstates, tol=None):
+
+    if nstates < 2:
+        raise ValueError(f"Number of bosonic states cannot be less than 2, provided {nstates}.")
+
+    cr_op = np.zeros((nstates, nstates))
+    for i in range(nstates - 1):
+        cr_op[i + 1, i] = np.sqrt(i + 1.0)
+
+    op_mat = {"+": cr_op, "-": cr_op.T}
```
Comment (austingmhuang) [inline pennylane/qchem/vibrational/bosonic_mapping.py:111]: Same comment as for binary mapping. Naming could be improved. I'm assuming the coefficients are calculated with some equation in the nature paper, but would be nice to have a comment to reference.

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    qubit_operator = PauliSentence({PauliWord({}): 1.0})
+
+    op_prod = defaultdict(list)
+
+    for (_, b_idx), sign in bose_operator.items():
```
Comment (austingmhuang) [inline pennylane/qchem/vibrational/bosonic_mapping.py:117]: Could do this in binary mapping pr as well.

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        op_prod[b_idx].append(sign)
```
Comment (austingmhuang) [inline pennylane/qchem/vibrational/bosonic_mapping.py:118]: What is op_prod? A `defaultdict` that tracks all the annihilation/creation operators acting on a site right? Is there a better name for this?

File: pennylane/qchem/vibrational/bosonic_mapping.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    for b_idx, signs in op_prod.items():
+        op_mat_prod = np.eye(nstates)
+        for sign in signs:
+            op_mat_prod = np.dot(op_mat_prod, op_mat[sign])
```
Comment (austingmhuang) [inline pennylane/qchem/vibrational/bosonic_mapping.py:123]: Not sure what op_mat_prod is. A reference would be useful to have.

File: tests/qchem/vibrational_tests/test_unary_mapping.py [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+            ],
+        ),
+    ),
+]
+
+
+@pytest.mark.parametrize("bose_op, nstates, result", BOSE_WORDS_AND_OPS)
+def test_unary_mapping_boseword(bose_op, nstates, result):
+    """Test that the unary_mapping function returns the correct qubit operator."""
+    # convert BoseWord to PauliSentence and simplify
```
Comment (austingmhuang) [inline tests/qchem/vibrational_tests/test_unary_mapping.py:211]:
```suggestion
```

File: tests/qchem/vibrational_tests/test_unary_mapping.py [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+    qubit_op = unary_mapping(bose_op, nstates=nstates, ps=True)
+    qubit_op.simplify(tol=1e-8)
+
+    # get expected op as PauliSentence and simplify
```
Comment (austingmhuang) [inline tests/qchem/vibrational_tests/test_unary_mapping.py:215]:
```suggestion
```

File: tests/qchem/vibrational_tests/test_unary_mapping.py [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+    assert ps_op == PauliSentence({})
+
+    op = unary_mapping(op).simplify()
+    assert isinstance(op, qml.ops.SProd)
+    assert isinstance(op.base, I)
+    assert op.scalar == 0
+
+
+def test_error_is_raised_for_incompatible_type():
+    """Test that an error is raised in the input is not a BoseWord or BoseSentence"""
```
Comment (austingmhuang) [inline tests/qchem/vibrational_tests/test_unary_mapping.py:261]:
```suggestion
    """Test that an error is raised if the input is not a BoseWord or BoseSentence"""
```

**Review by austingmhuang — CHANGES_REQUESTED** [review]
Documentation things. The naming in the main mapping function is a bit hard to follow; perhaps a reference would be nice?

File: pennylane/qchem/vibrational/bosonic_mapping.py (hunk 8 above)
Comment (ddhawan11) [reply pennylane/qchem/vibrational/bosonic_mapping.py:123]: In the nature paper, you see there is a section on avoiding superfluous terms, where the matrices are multiplied before converting them to Pauli operators. This is done to avoid some extra terms which do not add to the value of the operator but make it expensive to calculate. 
Added a comment in the code.

File: pennylane/qchem/vibrational/bosonic_mapping.py (hunk 7 above)
Comment (ddhawan11) [reply pennylane/qchem/vibrational/bosonic_mapping.py:118]: Yes, it is the product of operators acting on the same site, how about ops_per_index or ops_per_site?

File: pennylane/qchem/__init__.py [hunk 12]
```diff
@@ -14,6 +14,8 @@
 """
 This submodule provides the functionality to perform quantum chemistry calculations.
 """
+
+from .vibrational import unary_mapping
```
Comment (ddhawan11) [inline pennylane/qchem/__init__.py:18]:
```suggestion
```

File: pennylane/qchem/__init__.py [hunk 13]
```diff
@@ -82,3 +82,4 @@
     taper_hf,
     taper_operation,
 )
+
```
Comment (ddhawan11) [inline pennylane/qchem/__init__.py:85]:
```suggestion
```

File: pennylane/bose/bosonic_mapping.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+            the Bose operator to qubit wires. If None, integers used to
+            label the bosonic states will be used as wire labels. Defaults to None.
+        tol (float): tolerance for discarding the imaginary part of the coefficients
+
+    Returns:
+        Union[PauliSentence, Operator]: a linear combination of qubit operators
+
+    **Example**
+
+    >>> w = qml.BoseWord()
```
Comment (DSGuala) [inline pennylane/bose/bosonic_mapping.py:67]: Missing the input to `BoseWord` here.
```suggestion
    >>> w = qml.BoseWord({(0, 0): "+"})
```

File: pennylane/bose/__init__.py [hunk 15]
```diff
@@ -1,3 +1,17 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 """A module containing utility functions and mappings for working with bosonic operators. """
 
-from .bosonic import BoseWord, BoseSentence
+from .bosonic import BoseSentence, BoseWord
+from .bosonic_mapping import unary_mapping
```
Comment (DSGuala) [inline pennylane/bose/__init__.py:17]: Something going on with the imports.
I get errors for `qml.unary_mapping`. This may be fixed when the binary mapping PR is merged as it imports `bose` into the `pennylane/__init__`

File: pennylane/bose/bosonic_mapping.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+    r"""Convert a bosonic operator to a qubit operator using the unary mapping.
+
+    The mapping procedure is described in `arXiv.1909.12847 <https://arxiv.org/abs/1909.12847>`_.
+
+    Args:
+        bose_operator(BoseWord, BoseSentence): the bosonic operator
+        n_states(int): maximum number of allowed bosonic states
+        ps (bool): Whether to return the result as a PauliSentence instead of an
+            operator. Defaults to False.
+        wire_map (dict): A dictionary defining how to map the states of
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply pennylane/bose/bosonic_mapping.py]: Do we want to change it to same for other PRs as well?

File: pennylane/bose/bosonic_mapping.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+
+    return PauliSentence({PauliWord({}): c1, PauliWord({qub_id: "Z"}): c2})
+
+
+def unary_mapping(
+    bose_operator: Union[BoseWord, BoseSentence],
+    n_states: int = 2,
+    ps: bool = False,
+    wire_map: dict = None,
+    tol: float = None,
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply pennylane/bose/bosonic_mapping.py:46]: We intend to create another PR for doc fix, this is a new module so doesn't show.

File: pennylane/bose/bosonic_mapping.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane/bose/bosonic_mapping.py]: Hey @DSGuala actually I corrected this to use proper punctuation in accordance with the [pennylane documentation](https://docs.pennylane.ai/en/stable/development/guide/documentation.html). Just to clarify, should we follow the guide or is the guide outdated?

>Multi-sentence argument descriptions, or longer sentence fragments with mid-sentence punctuation marks, should use full capitalization and punctuation:

File: pennylane/qchem/vibrational/bosonic_mapping.py (hunk 7 above)
Comment (austingmhuang) [reply pennylane/qchem/vibrational/bosonic_mapping.py]: Yes, please change the name

File: pennylane/bose/bosonic_mapping.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (DSGuala) [reply pennylane/bose/bosonic_mapping.py]: My mistake. We should follow the Pennylane documentation guidelines.

**Review by soranjh — APPROVED** [review]

File: pennylane/bose/bosonic_mapping.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+        n_states(int): maximum number of allowed bosonic states
+        ps (bool): Whether to return the result as a PauliSentence instead of an
+            operator. Defaults to False.
+        wire_map (dict): A dictionary defining how to map the states of
+            the Bose operator to qubit wires. If None, integers used to
+            label the bosonic states will be used as wire labels. Defaults to None.
+        tol (float): tolerance for discarding the imaginary part of the coefficients
+
+    Returns:
+        Union[PauliSentence, Operator]: A linear combination of qubit operators
```
Comment (austingmhuang) [inline pennylane/bose/bosonic_mapping.py:63]:
```suggestion
        Union[PauliSentence, Operator]: A linear combination of qubit operators.
```

**Review by austingmhuang — APPROVED** [review]
Looks good @ddhawan11.

File: pennylane/bose/bosonic_mapping.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+    **Example**
+
+    >>> w = qml.BoseWord({(0, 0): "+"})
+    >>> qml.unary_mapping(w, n_states=4)
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply pennylane/bose/bosonic_mapping.py]: Double check that this works. I think `qml.FermiWord` doesn't work so that's why I made this change.

[the remark this answers is not in the document]
Comment (ddhawan11) [reply pennylane/bose/bosonic_mapping.py]: This one works, though I have changed it to be consistent with Fermi

File: tests/bose/test_unary_mapping.py [hunk 20]
```diff
@@ -0,0 +1,637 @@
...
+        ValueError, match="Number of allowed bosonic states cannot be less than 2, provided 0."
+    ):
+        unary_mapping(bw, n_states=0)
+
+
+def test_error_is_raised_for_incompatible_type():
+    """Test that an error is raised if the input is not a BoseWord or BoseSentence"""
+
+    with pytest.raises(ValueError, match="bose_operator must be a BoseWord or BoseSentence"):
+        unary_mapping(X(0))
```
Comment (austingmhuang) [inline tests/bose/test_unary_mapping.py:521]: Also just realized this should be a TypeError instead. But it's not blocking just would be nice to fix before merge.
